connector/connect/server/src/main/scala/org/apache/spark/sql/connect/execution/ConnectProgressExecutionListener.scala

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connect.execution
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.scheduler.{SparkListener, SparkListenerJobEnd, SparkListenerJobStart, SparkListenerStageCompleted, SparkListenerTaskEnd}
+
+/**
+ * A listener that tracks the execution of jobs and stages for a given set of tags. This is used
+ * to track the progress of a job that is being executed through the connect API.
+ *
+ * The listener is instantiated once for the SparkConnectService and then used to track all the
+ * current query executions.
+ */
+private[connect] class ConnectProgressExecutionListener extends SparkListener with Logging {
```

Comment:
Could we reuse `AppStatusListener` ? That is used by non Spark Connect code path (see `ConsoleProgressBar`)

Reply:
The `ConsoleProgressBar` "randomly" shows the progress of up to three stages. The stages can be completely independent of the actual query execution of this Spark Session. This is the reason why I created a custom listener that only tracks Spark Connect queries for the whole SparkConnect service and collapses them by operation tag.